nisystemlink/clients/notebook/models/_execution.py

```diff
@@ -0,0 +1,180 @@
+from datetime import datetime
+from enum import Enum
+from typing import Any, Dict, Optional
+
+from nisystemlink.clients.core._uplink._json_model import JsonModel
+from pydantic import Field
+
+
+class SourceType(str, Enum):
+    """Source type of an execution"""
+
+    MANUAL = "MANUAL"
+
+    TRIGGERED = "TRIGGERED"
+
+
+class Source(JsonModel):
+    """An object that defines properties set by routine service"""
+
+    type: SourceType
+    """Source type of an execution"""
+
+    routine_id: Optional[str] = None
+    """ID of the routine that triggered the execution"""
+
+    event_id: Optional[str] = None
+    """Unique identifier of event that triggered the execution"""
+
+
+class ReportType(str, Enum):
+    """Available types for a report that is going to be generated."""
+
+    NO_REPORT = "NO_REPORT"
+
+    HTML = "HTML"
+
+    PDF = "PDF"
+
+
+class ReportSettings(JsonModel):
+    """A class that defines settings of the Report"""
+
+    format: ReportType
+    """Type for the report that is going to be generated."""
+
+    exclude_code: bool
+    """Boolean parameter that will define if the source code should be included in the report or not."""
+
+
+class ExecutionPriority(str, Enum):
+    """Execution priority. Can be one of Low, Medium or High."""
+
+    LOW = "LOW"
+
+    MEDIUM = "MEDIUM"
+
+    HIGH = "HIGH"
+
+
+class ExecutionResourceProfile(str, Enum):
+
+    DEFAULT = "DEFAULT"
+
+    LOW = "LOW"
+
+    MEDIUM = "MEDIUM"
+
+    HIGH = "HIGH"
+
+
+class ExecutionStatus(str, Enum):
+    """Status of an execution."""
+
+    IN_PROGRESS = "IN_PROGRESS"
+
+    QUEUED = "QUEUED"
+
+    FAILED = "FAILED"
+
+    SUCCEEDED = "SUCCEEDED"
+
+    CANCELED = "CANCELED"
+
+    TIMED_OUT = "TIMED_OUT"
+
+
+class ExecutionErrorCode(str, Enum):
+    """Execution error code."""
+
+    NO_ERROR = "NO_ERROR"
+
+    NOTEBOOK_ERROR = "NOTEBOOK_ERROR"
+
+    NOTEBOOK_TIMEOUT_ERROR = "NOTEBOOK_TIMEOUT_ERROR"
+
+    NOTEBOOK_NOT_FOUND_ERROR = "NOTEBOOK_NOT_FOUND_ERROR"
+
+    NOTEBOOK_RESULT_TOO_BIG_ERROR = "NOTEBOOK_RESULT_TOO_BIG_ERROR"
+
+    NOT_PUBLISHED_ERROR = "NOT_PUBLISHED_ERROR"
+
+    OUT_OF_MEMORY_ERROR = "OUT_OF_MEMORY_ERROR"
+
+    UNKNOWN_ERROR = "UNKNOWN_ERROR"
+
+    DEAD_KERNEL_ERROR = "DEAD_KERNEL_ERROR"
+
+    EXECUTION_COULD_NOT_BE_RETRIED = "EXECUTION_COULD_NOT_BE_RETRIED"
+
+
+class Execution(JsonModel):
+    """Information about an execution of a Jupyter notebook that has the cachedResult field added."""
+
+    id: str
```

Comment:
All of these should be typed as Optional since the query API supports projections
```suggestion
    id: Optional[str]
```
Same for all fields in this class not already marked optional.